lib/modules/manager/flux/extract.ts

```diff
@@ -70,55 +84,149 @@ function readManifest(content: string, file: string): FluxManifest | null {
-  context: FluxManifest[]
-): PackageDependency<FluxManagerData>[] | null {
-  const resourceManifests = context.filter(
-    (manifest) => manifest.kind === 'resource'
-  ) as ResourceFluxManifest[];
-  const repositories = resourceManifests.flatMap(
-    (manifest) => manifest.repositories
-  );
-  let res: PackageDependency<FluxManagerData>[] | null = null;
-  switch (manifest.kind) {
-    case 'system':
-      res = [
-        {
-          depName: 'fluxcd/flux2',
-          datasource: GithubReleasesDatasource.id,
-          currentValue: manifest.version,
-          managerData: {
-            components: manifest.components,
-          },
-        },
-      ];
-      break;
-    case 'resource':
-      res = manifest.releases.map((release) => {
+const githubUrlRegex = regEx(
+  /^(?:https:\/\/|git@)github\.com[/:](?<packageName>[^/]+\/[^/]+?)(?:\.git)?$/
+);
+const gitlabUrlRegex = regEx(
+  /^(?:https:\/\/|git@)gitlab\.com[/:](?<packageName>[^/]+\/[^/]+?)(?:\.git)?$/
+);
+const bitbucketUrlRegex = regEx(
+  /^(?:https:\/\/|git@)bitbucket\.org[/:](?<packageName>[^/]+\/[^/]+?)(?:\.git)?$/
+);
+
+function resolveGitRepositoryPerSourceTag(
+  dep: PackageDependency,
+  gitUrl: string
+): void {
+  const githubMatchGroups = githubUrlRegex.exec(gitUrl)?.groups;
+  if (githubMatchGroups) {
+    dep.datasource = GithubTagsDatasource.id;
+    dep.packageName = githubMatchGroups.packageName;
+    dep.sourceUrl = `https://github.com/${dep.packageName}`;
+    return;
+  }
+
+  const gitlabMatchGroups = gitlabUrlRegex.exec(gitUrl)?.groups;
+  if (gitlabMatchGroups) {
+    dep.datasource = GitlabTagsDatasource.id;
+    dep.packageName = gitlabMatchGroups.packageName;
+    dep.sourceUrl = `https://gitlab.com/${dep.packageName}`;
+    return;
+  }
+
+  const bitbucketMatchGroups = bitbucketUrlRegex.exec(gitUrl)?.groups;
+  if (bitbucketMatchGroups) {
+    dep.datasource = BitBucketTagsDatasource.id;
+    dep.packageName = bitbucketMatchGroups.packageName;
+    dep.sourceUrl = `https://bitbucket.org/${dep.packageName}`;
+    return;
+  }
+
+  dep.datasource = GitTagsDatasource.id;
+  dep.packageName = gitUrl;
+  if (gitUrl.startsWith('https://')) {
+    dep.sourceUrl = gitUrl.replace(/\.git$/, '');
+  }
+}
+
+function resolveSystemManifest(
+  manifest: SystemFluxManifest
+): PackageDependency<FluxManagerData>[] {
+  return [
+    {
+      depName: 'fluxcd/flux2',
+      datasource: GithubReleasesDatasource.id,
+      currentValue: manifest.version,
+      managerData: {
+        components: manifest.components,
+      },
+    },
+  ];
+}
+
+function resolveResourceManifest(
+  manifest: ResourceFluxManifest,
+  context: ResourceFluxManifest[]
+): PackageDependency<FluxManagerData>[] {
+  const helmRepositories: HelmRepository[] = [];
+  for (const manifest of context) {
+    for (const resource of manifest.resources) {
+      if (resource.kind === 'HelmRepository') {
+        helmRepositories.push(resource);
+      }
+    }
+  }
```

Comment:
this should be moved one level up, so we don't do that loop for every manifest again.

Reply:
Do you mean that I will remove the `context` variable and replace it with `helmrepositories` then create the `helmrepositories` array in the `extractAllPackageFiles` function and pass it down via the parameters `extractAllPackageFiles->resolveManifest->resolveResourceManifest`?

Because the loop for the manifests happen in `extractAllPackageFiles`.
And the `resolveManifest` does handle both system and resource manifests, so passing down `helm` resources specifically from there while system resolve might not need them might be an anti-pattern, although i dont know the convention of typescript so im not sure.

What do you think?

Reply:
yes, but inline `resolveManifest` to `extractAllPackageFiles`, so we don't need that small extra function